quantum/process_keycode/process_joystick.c

```diff
@@ -0,0 +1,160 @@
+  } else {
+    if (record->event.pressed){
+      joystick_status.buttons[(keycode-JS_BUTTON0)/8] |= 1<<(keycode%8);
+    } else {
+      joystick_status.buttons[(keycode-JS_BUTTON0)/8] &= ~(1<<(keycode%8));
+    }
+    
+    joystick_status.status |= JS_UPDATED;
+  }
+  
+  return true;
+}
+
+uint8_t savePinState(uint8_t pin){
+#ifdef __AVR__
+  uint8_t pinNumber = pin & 0xF;
+  return ((PORTx_ADDRESS(pin) >> pinNumber) & 0x1) << 1 
+       | ((DDRx_ADDRESS(pin) >> pinNumber) & 0x1) ;
+#else
+  return 0;
+#endif
+}
+
+void restorePinState(uint8_t pin, uint8_t restoreState){
+#ifdef __AVR__
+  uint8_t pinNumber = pin & 0xF;
+  PORTx_ADDRESS(pin) = (PORTx_ADDRESS(pin) & ~_BV(pinNumber)) | (((restoreState >> 1) & 0x1) << pinNumber);
+  DDRx_ADDRESS(pin) = (DDRx_ADDRESS(pin) & ~_BV(pinNumber)) | ((restoreState & 0x1) << pinNumber);
+#else
+  return;
+#endif
+}
+
+__attribute__ ((weak))
+bool process_joystick_analogread(){
+  return process_joystick_analogread_quantum();
+}
+
+bool process_joystick_analogread_quantum(){
+
+#if JOYSTICK_AXES_COUNT > 0
+  for (int axis_index=0 ; axis_index<JOYSTICK_AXES_COUNT ; ++axis_index){
+    if (joystick_axes[axis_index].input_pin==JS_VIRTUAL_AXIS){
+      continue;
+    }
+    
+    //save previous input pin status as well
+    uint8_t inputSavedState = savePinState(joystick_axes[axis_index].input_pin);
+    
+    //disable pull-up resistor
+    writePinLow(joystick_axes[axis_index].input_pin);
+    
+    //if pin was a pull-up input, we need to uncharge it by turning it low 
+    // before making it a low input
+    setPinOutput(joystick_axes[axis_index].input_pin);
+    
+    wait_us(10);
+    
+    //save and apply output pin status
+    uint8_t outputSavedState = 0;
+    if (joystick_axes[axis_index].output_pin!=JS_VIRTUAL_AXIS) {
+      //save previous output pin status
+      outputSavedState = savePinState(joystick_axes[axis_index].output_pin);
+      
+      setPinOutput(joystick_axes[axis_index].output_pin);
+      writePinHigh(joystick_axes[axis_index].output_pin);
+    }
+    
+    uint8_t groundSavedState = 0;
+    if (joystick_axes[axis_index].ground_pin!=JS_VIRTUAL_AXIS) {
+      //save previous output pin status
+      groundSavedState = savePinState(joystick_axes[axis_index].ground_pin);
+      
+      setPinOutput(joystick_axes[axis_index].ground_pin);
+      writePinLow(joystick_axes[axis_index].ground_pin);
+    }
+    
+    wait_us(10);
+    
+    setPinInput(joystick_axes[axis_index].input_pin);
+    
+    wait_us(10);
+    
+#ifdef __AVR__
+    int16_t axis_val = analogReadPin(joystick_axes[axis_index].input_pin);
+#else
+  //default to resting position
+    int16_t axis_val = joystick_axes[axis_index].mid_digit;
+#endif
+    
+    //test the converted value against the lower range
+    uint16_t ref = joystick_axes[axis_index].mid_digit;
+    uint16_t range = joystick_axes[axis_index].min_digit;
+    int16_t ranged_val = -127*fminf(1.f, (axis_val - (float)(ref)) /(range - (float)ref));
```

Comment:
Is there any way we can avoid using floats? This will greatly increase the firmware size.

Reply:
I'll look into it